src/pip/_internal/vcs/subversion.py

```diff
@@ -29,6 +30,22 @@ class Subversion(VersionControl):
     repo_name = 'checkout'
     schemes = ('svn', 'svn+ssh', 'svn+http', 'svn+https', 'svn+svn')
 
+    def __init__(self, url=None, use_interactive=None, *args, **kwargs):
+        # type: (str, bool, *Any, **Any) -> None
+        if use_interactive is None:
+            use_interactive = sys.stdin.isatty()
+        self.use_interactive = use_interactive
+
+        # This member is used to cache the fetched version of the current
+        # ``svn`` client.
+        self._vcs_version = None  # type: Optional[Tuple[int, ...]]
+
+        # If the version of the ``svn`` client cannot be parsed, remember
+        # this to avoid extra unnecessary invocations of ``svn``.
+        self._call_vcs_version_failed = False
```

Comment:
One idea to make this all cleaner and do away with this extra instance attribute is to use the empty tuple `()` to signify the cannot-be-parsed case for both the  `get_vcs_version()` return value and the `_vcs_version` attribute value. I'm assuming that works with `Tuple[int, ...]`.

Reply:
A sum type (e.g. [Rust `enum`](https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html)) would be very useful here. 😄 Alas, Python's type system does not (yet) have such a concept.

```rust
enum FallibleVersion {
    CannotBeParsed,
    Parsed(i32, i32, i32)
}
```

But, I digress, yeah an empty tuple `()` would be one way to combine these states into a single variable. I'm generally wary of doing such things as now the type system can't encode this information. At the same time, `_call_vcs_version_failed` is not "attached" to `_vcs_version` in a type system way, so it has its own downsides. I do like that we can attach a meaningful name to it, though, and the type of `bool` is very specific and can only have two values.

So, I'm on the fence on this one. I agree the current way it's super beautiful, but encoding failure states into a certain value of the `tuple` also seems to have other drawbacks.

Reply:
It makes sense to me in that this way querying the version always gives you a tuple. It’s just that the empty tuple (falsey value) means “no version.” And None means not evaluated yet.

Reply:
Sounds good. I've update the PR to use this method.